src/platform/packages/shared/presentation/presentation_publishing/publishing_subject/publishing_batcher.ts

```diff
@@ -120,22 +120,37 @@ export const useBatchedPublishingSubjects = <
 
   /**
    * Subscribe to all subjects and update the latest values when any of them change.
+   *
+   * Can not set up subscription in useEffect.
+   * useEffect introduces a race condition where subscriptions may emit after values are set with useState
+   * but before subscription is setup in useEffect.
+   *
+   * Can not set up subscription in useRef.
+   * useRef executes initialization logic every render.
    */
-  useEffect(() => {
-    const subscription = combineLatest(subjects)
-      .pipe(
-        // When a new observer subscribes to a BehaviorSubject, it immediately receives the current value. Skip this emit.
-        skip(1),
-        debounceTime(0)
-      )
-      .subscribe((values) => {
-        setLatestPublishedValues(values as UnwrapPublishingSubjectTuple<SubjectsType>);
-      });
-    return () => subscription.unsubscribe();
+  const subscription = useMemo(
+    () =>
+      combineLatest(subjects)
+        .pipe(
+          // When a new observer subscribes to a BehaviorSubject, it immediately receives the current value. Skip this emit.
+          skip(1),
+          debounceTime(0)
+        )
+        .subscribe((values) => {
+          setLatestPublishedValues(values as UnwrapPublishingSubjectTuple<SubjectsType>);
+        }),
     // 'subjects' gets a new reference on each call because of spread
     // Use 'useBatchedOptionalPublishingSubjects' when 'subjects' are expected to change.
```

Comment:
I wonder if we should remove `useBatchedOptionalPublishingSubjects`. It _could_ be suffering from the same problem.

The reason it exists was mainly to allow state from an API to be consumed in the _same component_ where the API is rendered. The API is null for the first frame, so we can't extract the `publishingSubjects`.

If we removed `useBatchedOptionalPublishingSubjects`, in that edge case the consumer would have to wrap the JSX that consumes the API in another Component, and conditionally render that component based on the presence of the API. To me this seems like a more robust solution.

Reply:
As discussed
* Opened https://github.com/elastic/kibana/pull/216522 to fix race condition in `useStateFromPublishingSubject`
* Will open PR to remove `useBatchedOptionalPublishingSubjects`